src/Interpreters/TreeOptimizer.cpp

```diff
@@ -164,7 +168,7 @@ void optimizeGroupBy(ASTSelectQuery * select_query, const NameSet & source_colum
                     std::back_inserter(group_exprs), is_literal
             );
         }
-        else if (is_literal(group_exprs[i]))
+        else if (is_literal(group_exprs[i]) && !settings.enable_positional_arguments)
```

Comment:
I think when `enable_positional_arguments=true` we can also remove literals that cannot be indexes.